DOC more updates to maintainers' doc (#18954)
scikit-learn/scikit-learn: doc/developers/maintainer.rst

Preparing a release PR
......................

> Major version
> ~~~~~~~~~~~~~

Author: ```suggestion
Major version release
~~~~~~~~~~~~~~~~~~~~~
```

> Releasing the first RC of e.g. version `0.99` involves creating the release

Author: ```suggestion
Releasing the first RC of e.g. version `0.99.0` involves creating the release
```

contribute their changelog entries to this file in parallel to the release
process.

Minor version release
~~~~~~~~~~~~~~~~~~~~~

The minor releases should include bug fixes and some relevant documentation
changes only. Any PR resulting in a behavior change which is not a bug fix
should be excluded.

For major releases, do not forget to prepare a Release Highlights page as a
runnable example and check that its HTML rendering looks correct. These release
higlights should be linked from the ``doc/whats_new/v0.99.rst`` file for the
new version of scikit-learn.

First, create a branch, **on your own fork** (to release e.g. `0.99.3`):

.. prompt:: bash $

    # assuming master and upstream/master are the same
    git checkout -b release-0.99.3 master

Then, create a PR **to the** `scikit-learn/0.99.X` **branch** (not to
master!) with all the desired changes:

.. prompt:: bash $

	git rebase -i upstream/0.99.2

Do not forget to add a commit updating ``sklearn.__version__``.

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
understand what's included.

.. _making_a_release:

Making a release
................

> 0. Ensure that you have checkouted the branch of the release PR as explained

Author: ```suggestion
0. Ensure that you have checked out the branch of the release PR as explained
```

> 0. Ensure that you have checkouted the branch of the release PR as explained
>    :ref:`preparing_a_release_pr` above.

Author: ```suggestion
0. Ensure that you have checked out the branch of the release PR as explained
   :ref:`preparing_a_release_pr` above.
```

Preparing a release PR
......................

Major version
~~~~~~~~~~~~~

Releasing the first RC of e.g. version `0.99` involves creating the release
branch `0.99.X` directly on the main repo, where `X` really is the letter X,
**not a placeholder**. The development for the major and minor releases of 0.99
should **also** happen under `0.99.X`. Each release (rc, major, or minor) is a
tag under that branch.

This is done only once, as the major and minor releases happen on the same
branch:

   .. prompt:: bash $

>      git checkout -b 0.99.X

Reviewer: The explanation for this command now misses the key point that the release manager needs to do it either under a different clone from the main repo, or under the same clone, but from and to `upstream` instead of `origin`.

Author: I expanded the commands to be more explicit.

in the description of the Pull Request to track progress.

This PR will be used to push commits related to the release as explained in
:ref:`making_a_release`.

You can also create a second PR from master and targeting master to increment
the ``__version__`` variable in `sklearn/__init__.py` to increment the dev
version. This means while we're in the release candidate period, the latest
stable is two versions behind the master branch, instead of one. In this PR
targeting master you should also include a new file for the matching version
under the ``doc/whats_new/`` folder so PRs that target the next version can
contribute their changelog entries to this file in parallel to the release
process.

Minor version release
~~~~~~~~~~~~~~~~~~~~~

The minor releases should include bug fixes and some relevant documentation
changes only. Any PR resulting in a behavior change which is not a bug fix
should be excluded.

For major releases, do not forget to prepare a Release Highlights page as a
runnable example and check that its HTML rendering looks correct. These release
higlights should be linked from the ``doc/whats_new/v0.99.rst`` file for the

> new version of scikit-learn.

Reviewer: shouldn't this be under major release?

Author: indeed...

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
understand what's included.

.. _making_a_release:

Making a release
................

0. Ensure that you have checkouted the branch of the release PR as explained
   :ref:`preparing_a_release_pr` above.

1. Update docs. Note that this is for the final release, not necessarily for
   the RC releases. These changes should be made in master and cherry-picked
   into the release branch, only before the final release.

   - Edit the ``doc/whats_new/v0.99.rst`` file to add release title and list of
     contributors.
     You can retrieve the list of contributor names with:

     ::

       $ git shortlog -s 0.98.33.. | cut -f2- | sort --ignore-case | tr '\n' ';' | sed 's/;/, /g;s/, $//' | fold -s

     - For major releases, link the release highlights example from the ``doc/whats_new/v0.99.rst`` file.

   - Update the release date in ``whats_new.rst``

   - Edit the ``doc/templates/index.html`` to change the 'News' entry of the
     front page (with the release month as well).

2. On the branch for releasing, update the version number in
   ``sklearn/__init__.py``, the ``__version__``.

>    For major releases, please add a 0 add the end: 0.99.0 instead of 0.99.

Reviewer 2: ```suggestion
   For major releases, please add a 0 at the end: 0.99.0 instead of 0.99.
```

>    For a first release candidates, use the rc1 suffix on the expected final

Reviewer 2: Maybe:

```suggestion
   For the first release candidate, use the rc1 suffix on the expected final
```

Preparing a release PR
......................

Major version release
~~~~~~~~~~~~~~~~~~~~~

Prior to branching please do not forget to prepare a Release Highlights page as
a runnable example and check that its HTML rendering looks correct. These

> release higlights should be linked from the ``doc/whats_new/v0.99.rst`` file

Reviewer 2: ```suggestion
release highlights should be linked from the ``doc/whats_new/v0.99.rst`` file
```

for the new version of scikit-learn.

Releasing the first RC of e.g. version `0.99.0` involves creating the release
branch `0.99.X` directly on the main repo, where `X` really is the letter X,

> **not a placeholder**. The development for the major and minor releases of 0.99

Reviewer 2: ```suggestion
**not a placeholder**. The development for the major and minor releases of `0.99`
```

should **also** happen under `0.99.X`. Each release (rc, major, or minor) is a
tag under that branch.

This is done only once, as the major and minor releases happen on the same
branch:

   .. prompt:: bash $

     # Assuming upstream is an alias for the main scikit-learn repo:

>      git fetch uptream master

Reviewer 2: ```suggestion
     git fetch upstream master
```

>      git checkout uptream/master

Reviewer 2: ```suggestion
     git checkout upstream/master
```

     git checkout -b 0.99.X

>      git push --set-upstream uptream 0.99.X

Reviewer 2: ```suggestion
     git push --set-upstream upstream 0.99.X
```

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
understand what's included.

.. _making_a_release:

Making a release
................

0. Ensure that you have checked out the branch of the release PR as explained
   in :ref:`preparing_a_release_pr` above.

1. Update docs. Note that this is for the final release, not necessarily for
   the RC releases. These changes should be made in master and cherry-picked
   into the release branch, only before the final release.

   - Edit the ``doc/whats_new/v0.99.rst`` file to add release title and list of
     contributors.
     You can retrieve the list of contributor names with:

     ::

       $ git shortlog -s 0.98.33.. | cut -f2- | sort --ignore-case | tr '\n' ';' | sed 's/;/, /g;s/, $//' | fold -s

     - For major releases, link the release highlights example from the ``doc/whats_new/v0.99.rst`` file.

   - Update the release date in ``whats_new.rst``

   - Edit the ``doc/templates/index.html`` to change the 'News' entry of the
     front page (with the release month as well).

2. On the branch for releasing, update the version number in
   ``sklearn/__init__.py``, the ``__version__``.

>    For major releases, please add a 0 at the end: 0.99.0 instead of 0.99.

Reviewer 2: Just to ensure they all look the same:

```suggestion
   For major releases, please add a 0 at the end: `0.99.0` instead of `0.99`.
```

>    For the first release candidate, use the rc1 suffix on the expected final
>    release number: 0.99.0rc1.

Reviewer 2: Same here:

```suggestion
   For the first release candidate, use the `rc1` suffix on the expected final
   release number: `0.99.0rc1`.
```

In terms of including changes, the first RC ideally counts as a *feature
freeze*. Each coming release candidate and the final release afterwards will
include only minor documentation changes and bug fixes. Any major enhancement
or feature should be excluded.

The minor releases should include bug fixes and some relevant documentation
changes only. Any PR resulting in a behavior change which is not a bug fix
should be excluded.

For major releases, do not forget to prepare a Release Highlights page as a
runnable example and check that its HTML rendering looks correct. These release

> higlights should be linked from the ``whats_new.html`` file for the new version

Author: ```suggestion
higlights should be linked from the ``whats_new.rst`` file for the new version
```

> higlights should be linked from the ``whats_new.rst`` file for the new version
> of scikit-learn.

Reviewer 3: ```suggestion
higlights should be linked from the ``doc/whats_new/v0.99.rst`` file for the new version
of scikit-learn.
```

Reviewer 3: The [what's new page](https://scikit-learn.org/stable/whats_new.html) does not contain the link I think..

First, create a branch, **on your own fork** (to release e.g. `0.99.3`):

.. prompt:: bash $

    # assuming master and upstream/master are the same
    git checkout -b release-0.99.3 master

Then, create a PR **to the** `scikit-learn/0.99.X` **branch** (not to
master!) with all the desired changes:

.. prompt:: bash $

	git rebase -i upstream/0.99.2

> Do not forget to add a commit updating sklearn.__version__.

Reviewer 2: ```suggestion
Do not forget to add a commit updating ``sklearn.__version__``.
```

   the RC releases. These changes should be made in master and cherry-picked
   into the release branch, only before the final release.

   - Edit the ``doc/v0.99/whats_new.rst`` file to add release title and list of
     contributors.
     You can retrieve the list of contributor names with:

     ::

>        $ git shortlog -s 0.98.33.. | cut -f2- | sort --ignore-case | tr '\n' ';' | sed 's/;/, /g;s/, $//'

Reviewer 3: @ogrisel as I'm doing this right now may I propose this small change?
```suggestion
       $ git shortlog -s 0.98.33.. | cut -f2- | sort --ignore-case | tr '\n' ';' | sed 's/;/, /g;s/, $//' | fold -s
```
This allow to format the output with `n_char <= 80` and breaking at spaces

     - For major releases, link the release highlights example from the ``doc/whats_new/v0.99.rst`` file.

>    - Update the release date in ``whats_new.rst``

Reviewer 2: I do not see a release date in the `whats_new.rst` file 😕.

Author: We forgot on some of them in the past.

   the RC releases. These changes should be made in master and cherry-picked
   into the release branch, only before the final release.

   - Edit the ``doc/v0.99/whats_new.rst`` file to add release title and list of
     contributors.
     You can retrieve the list of contributor names with:

     ::

       $ git shortlog -s 0.98.33.. | cut -f2- | sort --ignore-case | tr '\n' ';' | sed 's/;/, /g;s/, $//'

>      For major releases, make sure that the release highlights example is
>      properly linked from the ``doc/v0.99/whats_new.rst`` file.

Reviewer 3: Can we say 
```suggestion
     - For major releases, link the release highlights example from the ``doc/v0.99/whats_new.rst`` file.
```
and remove this from above?

   The wheel building workflow is managed by GitHub Actions and the results be browsed at:
   https://github.com/scikit-learn/scikit-learn/actions?query=workflow%3A%22Wheel+builder%22

.. note::

  Before building the wheels, make sure that the ``pyproject.toml`` file is
  up to date and using the oldest version of ``numpy`` for each Python version
  to avoid `ABI <https://en.wikipedia.org/wiki/Application_binary_interface>`_
  incompatibility issues. Moreover, a new line have to be included in the

>   ``pyproject.toml`` file for each new supported version of Python.

Reviewer 2: I think that this information is not necessary if we decided to move on with `oldest-supported-numpy` #18900.

> 

Reviewer 2: I think that we have to add a new step here (before step 4) to mention that the wheels should be uploaded to the staging area using the "Run workflow" form for the `Wheel builder` workflow.

Nevertheless, if the PR from step 3 is merged to the release branch, this step would not be necessary (i.f.f. the commit message contains `[cd build]` marker).

However, I do not really like the idea of merging an empty PR to the release brach.

   The wheel building workflow is managed by GitHub Actions and the results be browsed at:
   https://github.com/scikit-learn/scikit-learn/actions?query=workflow%3A%22Wheel+builder%22

.. note::

  Before building the wheels, make sure that the ``pyproject.toml`` file is
  up to date and using the oldest version of ``numpy`` for each Python version

>   to avoid ABI incompatibility issues. Moreover, a new line have to be included

Reviewer 3: ```suggestion
  to avoid API incompatibility issues. Moreover, a new line have to be included
```

Author: This is ABI for Application Binary Interface. I will make that more explicit.

>   to avoid ABI incompatibility issues. Moreover, a new line have to be included
>   in the ``pyproject.toml`` file for each new supported version of Python.

Author: ```suggestion
  to avoid `ABI <https://en.wikipedia.org/wiki/Application_binary_interface>`_
  incompatibility issues. Moreover, a new line have to be included in the
  ``pyproject.toml`` file for each new supported version of Python.
```

Reviewer 3: Thanks!

> 4. Once the CD has completed successfully, upload the generated artifacts

Reviewer 3: CD? Perhaps CI?
Or maybe you could explain what CD means? At least I don't know what it is... :(

Author: https://en.wikipedia.org/wiki/Continuous_delivery

I will rephrase this paragraph and the previous one.

Reviewer 2: Sure @cmarmo, let me explain carefully.

CD means Continuous Delivery. It is the automated process by which the code is built (wheel), test, and then put into a staging or pre-production environment (Anaconda cloud).

Reviewer 2: Should we put this information in the maintainers' guide or a link to a proper definition should be enough?

   (.tar.gz and .whl files) to https://test.pypi.org using the "Run workflow"
   form for the following GitHub Actions workflow:

   https://github.com/scikit-learn/scikit-learn/actions?query=workflow%3A%22Publish+to+Pypi%22

5. If this went fine, you can proceed with tagging. Create the tag and push it (if
   it's an RC, it can be ``0.xxrc1`` for instance):

   .. prompt:: bash $

     git tag -a 0.99  # in the 0.99.X branch
     git push git@github.com:scikit-learn/scikit-learn.git 0.99

6. Trigger the GitHub Actions workflow again but this time to upload the artifacts
   to the real https://pypi.org (replace "testpypi" by "pypi" in the "Run
   workflow" form).

7. Alternatively, it's possible to collect locally the generated binary wheel

>    packages and source tarbakk and upload them all to PyPI by running the

Reviewer: ```suggestion
   packages and source tarball and upload them all to PyPI by running the
```

   The wheel building workflow is managed by GitHub Actions and the results be browsed at:
   https://github.com/scikit-learn/scikit-learn/actions?query=workflow%3A%22Wheel+builder%22

.. note::

  Before building the wheels, make sure that the ``pyproject.toml`` file is
  up to date and using the oldest version of ``numpy`` for each Python version
  to avoid `ABI <https://en.wikipedia.org/wiki/Application_binary_interface>`_
  incompatibility issues. Moreover, a new line have to be included in the
  ``pyproject.toml`` file for each new supported version of Python.

.. note::

  The acronym CD in `[cd build]` stands for `Continuous Delivery
  <https://en.wikipedia.org/wiki/Continuous_delivery>`_ and refers to the
  automation used to generate the release artifacts (binary and source
  packages). This can be seen as an extension to CI which stands for
  `Continuous Integration
  <https://en.wikipedia.org/wiki/Continuous_integration>`_. The CD workflow on

>   GitHub Actions is also used to automatically build and publish packages for
>   the developement branch of scikit-learn every night. See

Reviewer 2: ```suggestion
  GitHub Actions is also used to automatically create nightly builds and
  publish packages for the developement branch of scikit-learn. See
```
